watcher/local/watcher.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func handleStatesFromClient(currentTx channel.Transaction, statesSub statesSub, requestLatestTxn chan struct{},
+	latestTx chan channel.Transaction) {
+	var _tx channel.Transaction
+	var ok bool
+	for {
+		select {
+		case _tx, ok = <-statesSub.statesStream():
+			if !ok {
+				log.WithField("ID", currentTx.State.ID).Info("States sub closed by client. Shutting down handler")
+				return
+			}
+			currentTx = _tx
+			log.WithField("ID", currentTx.ID).Debugf("Received state from client", currentTx.Version, currentTx.ID)
+		case <-requestLatestTxn:
+			currentTx = receiveTxUntil(statesSub, time.NewTimer(statesFromClientWaitTime).C, currentTx)
+			latestTx <- currentTx
+		}
+	}
+}
+
+// receiveTxUntil wait for the transactions on statesPub until the timeout channel is closed
+// or statesPub is closed and returns the last received transaction.
+// If no transaction was received, then returns currentTx itself.
+func receiveTxUntil(statesSub statesSub, timeout <-chan time.Time, currentTx channel.Transaction) channel.Transaction {
+	var _tx channel.Transaction
+	var ok bool
+	for {
+		select {
+		case _tx, ok = <-statesSub.statesStream():
+			if !ok {
+				return currentTx // states sub was closed, send the latest event.
+			}
+			currentTx = _tx
+			log.WithField("ID", currentTx.ID).Debugf("Received state from client", currentTx.Version, currentTx.ID)
+		case <-timeout:
+			return currentTx // timer expired, send the latest the event.
+		}
+	}
+}
+
+func (w *Watcher) handleEventsFromChain(eventsFromChainSub channel.AdjudicatorSubscription,
+	eventsToClientPubSub adjudicatorPub, thisCh *ch) {
+	parent := thisCh
+	if thisCh.parent != nil {
+		parent = thisCh.parent
+	}
+
+	for e := eventsFromChainSub.Next(); e != nil; e = eventsFromChainSub.Next() {
+		switch e.(type) {
+		case *channel.RegisteredEvent:
+			parent.subChsAccess.Lock()
+			func() {
+				defer parent.subChsAccess.Unlock()
+
+				log := log.WithFields(log.Fields{"ID": e.ID(), "Version": e.Version()})
+				log.Debug("Received registered event from chain")
+
+				eventsToClientPubSub.publish(e)
+
+				latestTx := thisCh.retreiveLatestTx()
+				log.Debugf("Latest version is (%d)", latestTx.Version)
+
+				if e.Version() < latestTx.Version {
+					if e.Version() < thisCh.registeredVersion {
+						log.Debugf("Latest version (%d) already registered ", thisCh.registeredVersion)
+						return
+					}
+					log.Debugf("Registering latest version (%d)", latestTx.Version)
+					err := registerDispute(w.registry, w.rs, parent)
+					if err != nil {
+						log.Error("Error registering dispute")
+						// TODO: Should the subscription be closed ?
+						return
+					}
+					log.Debug("Registered successfully")
+				}
+			}()
+		case *channel.ProgressedEvent:
+			log.Debugf("Received progressed event from chain: %v", e)
+			eventsToClientPubSub.publish(e)
+		case *channel.ConcludedEvent:
+			log.Debugf("Received concluded event from chain: %v", e)
+			eventsToClientPubSub.publish(e)
+		default:
```

Comment:
Could you also log this case, that an event was ignored?

Reply:
Log.Warn or Error that an unknown event type was observed. This should never happen.

Reply:
Logged as error. See 5088c4c.